tests/integration_tests/test_upgrade.py

```diff
@@ -160,13 +162,65 @@ def test_cosmovisor_upgrade(custom_ethermint: Ethermint):
     )
     assert receipt.status == 1
 
-    # check json-rpc query on older blocks works
-    assert old_balance == w3.eth.get_balance(
-        ADDRS["validator"], block_identifier=old_height
+    # check get_balance and eth_call don't work on pruned state
+    with pytest.raises(Exception):
+        w3.eth.get_balance(validator, block_identifier=old_height)
+
+    base_dir = custom_ethermint.base_dir
+    for i in [0, 1]:
+        supervisorctl(base_dir / "../tasks.ini", "stop", f"ethermint_9000-1-node{i}")
+
+    procs = []
+
+    def append_proc(log, cmd):
+        with (base_dir / log).open("a") as logfile:
+            procs.append(
+                subprocess.Popen(
+                    cmd,
+                    stdout=logfile,
+                    stderr=subprocess.STDOUT,
+                )
+            )
+
+    path = Path(custom_ethermint.chain_binary).parent.parent.parent
+    append_proc(
+        "node1.log",
+        [
+            f"{str(path)}/genesis/bin/ethermintd",
+            "start",
+            "--home",
+            base_dir / "node1",
+        ],
+    )
+    grpc_port1 = ports.grpc_port(custom_ethermint.base_port(1))
+    append_proc(
+        "node0.log",
+        [
+            f"{str(path)}/{plan_name}/bin/ethermintd",
+            "start",
+            "--json-rpc.backup-grpc-address-block-range",
+            f'{{"0.0.0.0:{grpc_port1}": [0, {target_height}]}}',
```

Comment:
@yihuang any better way to avoid this monstrosity, need support config like
```
'backup-grpc-address-block-range': {
  '0.0.0.0:26113': [0, 20],
  '0.0.0.0:26114': [20, 30],
}
```

Reply:
either a custom format string or json looks good to me